### ISSUE TITLE
Using Boost Iostreams on Windows with options does not link correctly

### ISSUE THREAD

**Issue by aadityakalsi** [ruslo/hunter#139]
Using Hunter with Boost on Windows as directed by the pkg wiki:

```
set(Boost_USE_STATIC_LIBS ON)
if(MSVC)
  add_definitions(-DBOOST_ALL_NO_LIB=1)
endif()

```

This leads to compilation success but NOT link success:

```
CustomBuild:
  Building Custom Rule C:/Users/Aaditya/projects/minitar/CMakeLists.txt
  CMake does not need to re-run because C:\Users\Aaditya\projects\minitar\build\CMakeFiles\generate.stamp is up-to-date.
ClCompile:
  C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\bin\CL.exe /c /IC:\Users\Aaditya\projects /Ic:\Users\Aaditya\.hunter\_Base\8438f64\a6a753a\4d0397b\Install\include /Zi /nologo /W3 /WX /Od /Ob0 /Oy- /D WIN32 /D _WINDOWS /D _DEBUG /D BOOST_ALL_NO_LIB=1 /D _CRT_SECURE_NO_WARNINGS /D _SCL_SECURE_NO_WARNINGS /D "CMAKE_INTDIR=\"Debug\"" /D _MBCS /Gm- /EHsc /RTC1 /MDd /GS /fp:precise /Zc:wchar_t /Zc:forScope /GR /Fo"main.dir\Debug\\" /Fd"main.dir\Debug\vc120.pdb" /Gd /TP /analyze- /errorReport:queue C:\Users\Aaditya\projects\minitar\main.cpp C:\Users\Aaditya\projects\minitar\headers.cpp C:\Users\Aaditya\projects\minitar\path.cpp C:\Users\Aaditya\projects\minitar\streams.cpp
  main.cpp
  headers.cpp
  path.cpp
  streams.cpp
  Generating Code...
Link:
  C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\bin\link.exe /ERRORREPORT:QUEUE /OUT:"C:\Users\Aaditya\projects\minitar\build\Debug\main.exe" /INCREMENTAL /NOLOGO kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /DEBUG /PDB:"C:/Users/Aaditya/projects/minitar/build/Debug/main.pdb" /SUBSYSTEM:CONSOLE /TLBID:1 /DYNAMICBASE /NXCOMPAT /IMPLIB:"C:/Users/Aaditya/projects/minitar/build/Debug/main.lib" /MACHINE:X86 /SAFESEH  /machine:X86 /debug c:/Users/Aaditya/.hunter/_Base/8438f64/a6a753a/4d0397b/Install/lib/zlib.lib c:/Users/Aaditya/.hunter/_Base/8438f64/a6a753a/4d0397b/Install/lib/bz2.lib c:/Users/Aaditya/.hunter/_Base/8438f64/a6a753a/4d0397b/Install/lib/libboost_filesystem-mt.lib c:/Users/Aaditya/.hunter/_Base/8438f64/a6a753a/4d0397b/Install/lib/libboost_iostreams-mt.lib c:/Users/Aaditya/.hunter/_Base/8438f64/a6a753a/4d0397b/Install/lib/libboost_system-mt.lib main.dir\Debug\main.obj
  main.dir\Debug\headers.obj
  main.dir\Debug\path.obj
  main.dir\Debug\streams.obj
streams.obj : error LNK2019: unresolved external symbol "public: static void __cdecl boost::iostreams::zlib_error::check(int)" (?check@zlib_error@iostreams@boost@@SAXH@Z) referenced in function "public: bool __thiscall boost::iostreams::detail::zlib_compressor_impl<class std::allocator<char> >::filter(char const * &,char const *,char * &,char *,bool)" (?filter@?$zlib_compressor_impl@V?$allocator@D@std@@@detail@iostreams@boost@@QAE_NAAPBDPBDAAPADPAD_N@Z) [C:\Users\Aaditya\projects\minitar\build\main.vcxproj]
streams.obj : error LNK2019: unresolved external symbol "protected: __thiscall boost::iostreams::detail::zlib_base::zlib_base(void)" (??0zlib_base@detail@iostreams@boost@@IAE@XZ) referenced in function "public: __thiscall boost::iostreams::detail::zlib_compressor_impl<class std::allocator<char> >::zlib_compressor_impl<class std::allocator<char> >(struct boost::iostreams::zlib_params const &)" (??0?$zlib_compressor_impl@V?$allocator@D@std@@@detail@iostreams@boost@@QAE@ABUzlib_params@23@@Z) [C:\Users\Aaditya\projects\minitar\build\main.vcxproj]
streams.obj : error LNK2019: unresolved external symbol "protected: __thiscall boost::iostreams::detail::zlib_base::~zlib_base(void)" (??1zlib_base@detail@iostreams@boost@@IAE@XZ) referenced in function __unwindfunclet$??0?$zlib_compressor_impl@V?$allocator@D@std@@@detail@iostreams@boost@@QAE@ABUzlib_params@23@@Z$0 [C:\Users\Aaditya\projects\minitar\build\main.vcxproj]
streams.obj : error LNK2019: unresolved external symbol "protected: void __thiscall boost::iostreams::detail::zlib_base::before(char const * &,char const *,char * &,char *)" (?before@zlib_base@detail@iostreams@boost@@IAEXAAPBDPBDAAPADPAD@Z) referenced in function "public: bool __thiscall boost::iostreams::detail::zlib_compressor_impl<class std::allocator<char> >::filter(char const * &,char const *,char * &,char *,bool)" (?filter@?$zlib_compressor_impl@V?$allocator@D@std@@@detail@iostreams@boost@@QAE_NAAPBDPBDAAPADPAD_N@Z) [C:\Users\Aaditya\projects\minitar\build\main.vcxproj]
streams.obj : error LNK2019: unresolved external symbol "protected: void __thiscall boost::iostreams::detail::zlib_base::after(char const * &,char * &,bool)" (?after@zlib_base@detail@iostreams@boost@@IAEXAAPBDAAPAD_N@Z) referenced in function "public: bool __thiscall boost::iostreams::detail::zlib_compressor_impl<class std::allocator<char> >::filter(char const * &,char const *,char * &,char *,bool)" (?filter@?$zlib_compressor_impl@V?$allocator@D@std@@@detail@iostreams@boost@@QAE_NAAPBDPBDAAPADPAD_N@Z) [C:\Users\Aaditya\projects\minitar\build\main.vcxproj]
```

The same build works correctly on Linux/Mac

**Comment by aadityakalsi** [comment]
Turns out, this could be related to the default '-s' arguments on Windows.

This is a bug in ```hunter_boost_component_b2_args``` where on Windows, it should use Windows' defaults.

**Comment by ruslo** [comment]
I see debug variant, what about release?

**Comment by aadityakalsi** [comment]
I was trying to link to my own code.

Turns out the issue wasa b2_component_args which was not accoounting for different defaults of Boost Iostreams on Windows.

I have a temporary testing space where I fixed that issue. but then I ran into this:

```
-- [hunter] [ Hunter-ID: 8438f64 | Config-ID: a6a753a | Toolchain-ID: 4d0397b ]

-- [hunter] ZLIB_ROOT: c:/Users/Aaditya/.hunter/_Base/8438f64/a6a753a/4d0397b/Install (ver.: 1.2.8-p3)

-- [hunter] BZIP2_ROOT: c:/Users/Aaditya/.hunter/_Base/8438f64/a6a753a/4d0397b/Install (ver.: 1.0.6-p1)

-- [hunter] Boost -s args: -s;NO_COMPRESSION=0;-s;NO_ZLIB=0;-s;NO_BZIP2=0;-s;ZLIB_INCLUDE=c:/Users/Aaditya/.hunter/_Base/8438f64/a6a753a/4d0397b/Install/include;-s;ZLIB_LIBPATH=c:/Users/Aaditya/.hunter/_Base/8438f64/a6a753a/4d0397b/Install/lib/zlib.lib;-s;BZIP2_INCLUDE=c:/Users/Aaditya/.hunter/_Base/8438f64/a6a753a/4d0397b/Install/include;-s;BZIP2_LIBPATH=c:/Users/Aaditya/.hunter/_Base/8438f64/a6a753a/4d0397b/Install/lib/bz2.lib

CMake Warning (dev) at c:/Program Files (x86)/CMake/share/cmake-3.2/Modules/ExternalProject.cmake:1815 (if):

  Policy CMP0054 is not set: Only interpret if() arguments as variables or

  keywords when unquoted.  Run "cmake --help-policy CMP0054" for policy

  details.  Use the cmake_policy command to set the policy and suppress this

  warning.



  Quoted variables like "x" will no longer be dereferenced when the policy is

  set to NEW.  Since the policy is not set the OLD behavior will be used.

Call Stack (most recent call first):

  c:/Program Files (x86)/CMake/share/cmake-3.2/Modules/ExternalProject.cmake:2321 (_ep_add_download_command)

  CMakeLists.txt:257 (ExternalProject_Add)

This warning is for project developers.  Use -Wno-dev to suppress it.



-- Configuring done

-- Generating done

-- Build files have been written to: c:/Users/Aaditya/.hunter/_Base/8438f64/a6a753a/4d0397b/Build/Boost/__iostreams/Build

Microsoft (R) Build Engine version 12.0.30723.0

[Microsoft .NET Framework, version 4.0.30319.34014]

Copyright (C) Microsoft Corporation. All rights reserved.



Build started 5/5/2015 7:57:02 PM.

Project "c:\Users\Aaditya\.hunter\_Base\8438f64\a6a753a\4d0397b\Build\Boost\__iostreams\Build\ALL_BUILD.vcxproj" on node 1 (default targets).

Project "c:\Users\Aaditya\.hunter\_Base\8438f64\a6a753a\4d0397b\Build\Boost\__iostreams\Build\ALL_BUILD.vcxproj" (1) is building "c:\Users\Aaditya\.hunter\_Base\8438f64\a6a753a\4d0397b\Build\Boost\__iostreams\Build\Boost-iostreams.vcxproj" (2) on node 1 (default targets).

Project "c:\Users\Aaditya\.hunter\_Base\8438f64\a6a753a\4d0397b\Build\Boost\__iostreams\Build\Boost-iostreams.vcxproj" (2) is building "c:\Users\Aaditya\.hunter\_Base\8438f64\a6a753a\4d0397b\Build\Boost\__iostreams\Build\ZERO_CHECK.vcxproj" (3) on node 1 (default targets).

PrepareForBuild:

  Creating directory "Win32\Debug\ZERO_CHECK\".

  Creating directory "c:\Users\Aaditya\.hunter\_Base\8438f64\a6a753a\4d0397b\Build\Boost\__iostreams\Build\Debug\".

  Creating directory "Win32\Debug\ZERO_CHECK\ZERO_CHECK.tlog\".

InitializeBuildStatus:

  Creating "Win32\Debug\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.

CustomBuild:

  Checking Build System

  CMake does not need to re-run because c:/Users/Aaditya/.hunter/_Base/8438f64/a6a753a/4d0397b/Build/Boost/__iostreams/Build/CMakeFiles/generate.stamp is up-to-date.

FinalizeBuildStatus:

  Deleting file "Win32\Debug\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild".

  Touching "Win32\Debug\ZERO_CHECK\ZERO_CHECK.tlog\ZERO_CHECK.lastbuildstate".

Done Building Project "c:\Users\Aaditya\.hunter\_Base\8438f64\a6a753a\4d0397b\Build\Boost\__iostreams\Build\ZERO_CHECK.vcxproj" (default targets).

PrepareForBuild:

  Creating directory "Win32\Debug\Boost-iostreams\".

  Creating directory "Win32\Debug\Boost-iostreams\Boost-iostreams.tlog\".

InitializeBuildStatus:

  Creating "Win32\Debug\Boost-iostreams\Boost-iostreams.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.

ComputeCustomBuildOutput:

  Creating directory "c:\Users\Aaditya\.hunter\_Base\8438f64\a6a753a\4d0397b\Build\Boost\__iostreams\Build\Boost-iostreams-prefix\src\Boost-iostreams-stamp\Debug\".

  Creating directory "c:\Users\Aaditya\.hunter\_Base\8438f64\a6a753a\4d0397b\Build\Boost\__iostreams\Build\CMakeFiles\Debug\".

CustomBuild:

  Building Custom Rule C:/Users/Aaditya/.hunter/_Base/8438f64/a6a753a/4d0397b/Build/Boost/__iostreams/CMakeLists.txt

  CMake does not need to re-run because c:\Users\Aaditya\.hunter\_Base\8438f64\a6a753a\4d0397b\Build\Boost\__iostreams\Build\CMakeFiles\generate.stamp is up-to-date.

  Creating directories for 'Boost-iostreams'

  Performing download step (download, verify and extract) for 'Boost-iostreams'

  -- verifying file...

       file='c:/Users/Aaditya/.hunter/_Base/Download/Boost/1.57.0/9305649/v1.57.0.tar.gz'

  -- verifying file... done

  -- extracting...

       src='c:/Users/Aaditya/.hunter/_Base/Download/Boost/1.57.0/9305649/v1.57.0.tar.gz'

       dst='c:/Users/Aaditya/.hunter/_Base/8438f64/a6a753a/4d0397b/Build/Boost/__iostreams/Source'

  -- extracting... [tar xfz]

  -- extracting... [analysis]

  -- extracting... [rename]

  -- extracting... [clean up]

  -- extracting... done

  No update step for 'Boost-iostreams'

  No patch step for 'Boost-iostreams'

  Performing configure step for 'Boost-iostreams'

  Building Boost.Build engine

  

  Bootstrapping is done. To build, run:

  

      .\b2

      

  To adjust configuration, edit 'project-config.jam'.

  Further information:

  

      - Command line help:

      .\b2 --help

       

      - Getting started guide: 

      http://boost.org/more/getting_started/windows.html

       

      - Boost.Build documentation:

      http://www.boost.org/boost-build2/doc/html/index.html

  Performing build step for 'Boost-iostreams'

  

  Building the Boost C++ Libraries.

  

  

CUSTOMBUILD : error : No best alternative for libs/iostreams/build/boost_bzip2 [c:\Users\Aaditya\.hunter\_Base\8438f64\a6a753a\4d0397b\Build\Boost\__iostreams\Build\Boost-iostreams.vcxproj]

      next alternative: required properties: (empty)

          matched

      next alternative: required properties: (empty)

          matched

  Performing configuration checks

  

      - zlib                     : no

CUSTOMBUILD : error : No best alternative for libs/iostreams/build/boost_bzip2 [c:\Users\Aaditya\.hunter\_Base\8438f64\a6a753a\4d0397b\Build\Boost\__iostreams\Build\Boost-iostreams.vcxproj]

      next alternative: required properties: (empty)

          matched

      next alternative: required properties: (empty)

          matched

CUSTOMBUILD : error : No best alternative for libs/iostreams/build/boost_bzip2 [c:\Users\Aaditya\.hunter\_Base\8438f64\a6a753a\4d0397b\Build\Boost\__iostreams\Build\Boost-iostreams.vcxproj]

      next alternative: required properties: (empty)

          matched

      next alternative: required properties: (empty)

          matched

      - zlib                     : no  (cached)

CUSTOMBUILD : error : No best alternative for libs/iostreams/build/boost_bzip2 [c:\Users\Aaditya\.hunter\_Base\8438f64\a6a753a\4d0397b\Build\Boost\__iostreams\Build\Boost-iostreams.vcxproj]

      next alternative: required properties: (empty)

          matched

      next alternative: required properties: (empty)

          matched
```
 And at compile time:

```
  compile-c-c++ bin.v2\libs\iostreams\build\msvc-12.0\debug\link-static\threading-multi\bzip2.obj

  bzip2.cpp

libs\iostreams\src\bzip2.cpp(20): fatal error C1083: Cannot open include file: 'bzlib.h': No such file or directory [c:\Users\Aaditya\.hunter\_Base\8438f64\a6a753a\4d0397b\Build\Boost\__iostreams\Build\Boost-iostreams.vcxproj]

  

      call "C:\Users\Aaditya\AppData\Local\Temp\b2_msvc_12.0_vcvarsall_x86.cmd" >nul

  cl /Zm800 -nologo @"bin.v2\libs\iostreams\build\msvc-12.0\debug\link-static\threading-multi\bzip2.obj.rsp" 

  

  ...failed compile-c-c++ bin.v2\libs\iostreams\build\msvc-12.0\debug\link-static\threading-multi\bzip2.obj...

  ...skipped <pbin.v2\libs\iostreams\build\msvc-12.0\debug\link-static\threading-multi>libboost_iostreams-mt-gd.lib for lack of <pbin.v2\libs\iostreams\build\msvc-12.0\debug\link-static\threading-multi>bzip2.obj...

  ...skipped <pstage\lib>libboost_iostreams-mt-gd.lib for lack of <pbin.v2\libs\iostreams\build\msvc-12.0\debug\link-static\threading-multi>libboost_iostreams-mt-gd.lib...
```

Seems like Boost does not find ZLIB and BZip2 despite the right -S settings. Do we have to use '/' instead of the Unix like '-' for options? Why are the ```b2``` arguments not being passed correctly or being miscalculated by boost?

**Comment by ruslo** [comment]
> -s;ZLIB_INCLUDE=c:/Users/Aaditya/.hunter/_Base/8438f64/a6a753a/4d0397b/Install/include;

probably boost expecting `-s;ZLIB_INCLUDE=c:\Users\Aaditya\.hunter\_Base\8438f64\a6a753a\4d0397b\Install\include`? You can try [file(TO_NATIVE_PATH ...)](http://www.cmake.org/cmake/help/v3.2/command/file.html) to fix it.

Can you please add an [example](https://github.com/ruslo/hunter/tree/master/examples) so we can test it in future.

**Comment by aadityakalsi** [comment]
Will do.. Still trying to get this to work.. Will add a patch and an example after asking you about how best to do it..

**Comment by aadityakalsi** [comment]
I still get the same error message. Using TO_NATIVE_PATH didn't help.

Modifying the ```verbose_output``` flag to be ON to see what's going in in Boost.

It seems like some incorrect flags, because there is nothing obvious that is going wrong.

**Comment by aadityakalsi** [comment]
So essentially, Boost's ```build_opts``` seem correct.

However, the boost b2 command runs with weird compile options.

Here is a gist with details:

https://gist.github.com/aadityakalsi/3c71abb57ecff81a297c

**Comment by ruslo** [comment]
> Will add a patch and an example after asking you about how best to do it

You can add new CMakeLists.txt to examples directory. May be also you need to modify [common.cmake](https://github.com/ruslo/hunter/blob/master/examples/common.cmake) file since we need extra build options:
```cmake
# common.cmake
HunterGate(... ${filepath_option})
```

```cmake
set(filepath_option FILEPATH "${CMAKE_CURRENT_LIST_DIR}/config.cmake")
include(../common.cmake) # emulate HunterGate
```
For other packages `filepath_option` will be empty, so will have no effect. + add this example to Travis config.

> Modifying the verbose_output flag to be ON to see what's going in in Boost

You can just add `HUNTER_STATUS_DEBUG=ON`

**Comment by ruslo** [comment]
> Here is a gist with details:

What about same run on Linux or Mac platform? Looks like this is boost issue, I see no `-I/path/to/bzip2/include` flag passed to compiler:
```
  file bin.v2\libs\iostreams\build\msvc-12.0\release\link-static\threading-multi\bzip2.obj.rsp
  "libs\iostreams\src\bzip2.cpp" -Fo"bin.v2\libs\iostreams\build\msvc-12.0\release\link-static\threading-multi\bzip2.obj"    -TP /O2 /Ob2 /W3 /GR /MD /Zc:forScope /Zc:wchar_t /wd4675 /EHs  /MP /MP /MP /MP /MP /MP /DBOOST_ALL_NO_LIB=1 /DNOMINMAX /DHAVE_ROUND -c 
  -DBOOST_ALL_NO_LIB=1 
  -DBOOST_IOSTREAMS_USE_DEPRECATED 
  -DNDEBUG 
  "-I."
  compile-c-c++ bin.v2\libs\iostreams\build\msvc-12.0\release\link-static\threading-multi\bzip2.obj
```

**Comment by aadityakalsi** [comment]
Exactly. It works correctly on both my Linux and Mac boxes.

How do you propose we debug this?

**Comment by ruslo** [comment]
> How do you propose we debug this?

Well it's interesting to see what is the log on Linux/OSX. Does it have "-I/path/to/zlib"?

> Can you tell me where the command to run during the Boost build is stored?

See `b2_cmd`, `build_opts` and `bootstrap_cmd` variables. Also everything will be printed (including user's jam file) if you will use `HUNTER_STATUS_DEBUG=ON`.

By the way does it work if you build boost with this options without Hunter? What if you build boost yourself with additional `-s ZLIB_INCLUDE=/path/to/hunter/install`?

**Comment by aadityakalsi** [comment]
Here is the OSX build log:
https://gist.github.com/aadityakalsi/1b79a0499cabe2cbeb6b

I haven't had enough time to try this. Will do this soon.

**Comment by ruslo** [comment]
Thanks for the log. If you grep bzip2.cpp you will find:
```
"/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++"  -ftemplate-depth-128 -O0 -fno-inline -Wall -g -gdwarf-2 -fexceptions  -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.9.sdk -DBOOST_ALL_NO_LIB=1 -DBOOST_IOSTREAMS_USE_DEPRECATED  -I"." -I"/Users/akalsi/.hunter/_Base/8438f64/fb721d1/373518b/Install/include" -c -o "bin.v2/libs/iostreams/build/darwin-4.2.1/debug/link-static/threading-multi/bzip2.o" "libs/iostreams/src/bzip2.cpp"
```

As you can see the `-I` option used to add location of header `bzlib.h`:
```
-I"/Users/akalsi/.hunter/_Base/8438f64/fb721d1/373518b/Install/include"
```

Same expected while building with Visual Studio compiler.

**Comment by aadityakalsi** [comment]
I have been able to get this to work on Windows. The issue was the default values of ```BZIP2_BINARY``` and ```ZLIB_BINARY```.

[SO post](http://stackoverflow.com/questions/7282645/how-to-build-boost-iostreams-with-gzip-and-bzip2-support-on-windows)

Patch coming soon. I also faced issues of setting variables inside ```ÌF``` blocks in CMake code.

May need your help cleaning those up and getting an example.
